### PR TITLE
[graph_trainer] in-graph inplace gradient accumulation mode, to make it CudaGraph friendly

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Option to include grad accumulation inside the graph and do inplace accumulation, which is CudaGraphable.

For this GradState is extracted to a dataclass,
Grad accumulation is included into a graph as add_ node.

The list of functionality is not supported yet and needs additional testing - there is extensive validation:
 - no parameters aliasing (tight params)
 - no PP
 - no Precompile

Authored with assistance from an AI coding assistant.

### CODE REVIEW

File: torchtitan/experiments/graph_trainer/trainer.py [hunk 1]
```diff
@@ -103,7 +104,27 @@ def fwd_bwd_step(inputs, labels, global_valid_tokens, extra_kwargs):
...
+
+        def fwd_bwd_step(inputs, labels, global_valid_tokens, extra_kwargs):
+            loss, _named_params, grads = compute_step(
+                inputs, labels, global_valid_tokens, extra_kwargs
+            )
+            return [loss, *grads]
+
+        return fwd_bwd_step
+
+    def fwd_bwd_step(
```
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py:119]: nice!

File: torchtitan/experiments/graph_trainer/trainer.py
Comment (aditvenk) [inline torchtitan/experiments/graph_trainer/trainer.py:300]: nit: The if and else clauses can be refactored to fewer lines.

**Review by aditvenk — APPROVED** [review]

File: torchtitan/trainer.py [hunk 2]
```diff
@@ -875,8 +875,12 @@ def _pp_forward_backward_body(
             return torch.sum(torch.stack(detached_losses)).to(self.device)
         return self._pp_loss_sentinel_on_non_last_stage
 
-    def train_step(self, data_iterator: Iterator[dict[str, Any]]):
+    def _zero_grad(self) -> None:
+        """Clear parameter gradients before one optimizer step."""
         self.optimizers.zero_grad(set_to_none=self.config.training.disable_cuda_graphs)
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: @fegin IIUC, you'll do Boyuan's improvement so that we will always `set_to_none=True`, is that correct?

@IvanKobzarev If that'd be the case, do you still need to override?

Comment (IvanKobzarev) [reply torchtitan/trainer.py]: Yes, we can avoid this, moving the logic to GraphTrainer.
This was mostly to have one explicit place where zero_grad happens.

Without it, It will be compatible with both set_to_none=False / True without deps on other changes, but will have slightly different paths.

set_to_none=False: Grad buffers are bound permanently to GraphState, optimizer.zero_grad will zero it.
set_to_none=True - gradients are in GraphGradientState.buffers, only param.grad ref is None, GraphGradientState.prepare_for_backward will zero it.

If we get to always set_to_none=True, this will be consolidation and easier to reason.
